### PR TITLE
feat(node-live-text-to-speech): use @deepgram/sdk for the live TTS connection

### PR DESCRIPTION
**Pull request by GregHolmes**
## What

Migrates the backend from a raw WebSocket proxy (to `wss://api.deepgram.com/v1/speak`) to the official [`@deepgram/sdk`](https://www.npmjs.com/package/@deepgram/sdk) (`^5.7.0`), using `client.speak.v1` streaming support for the Deepgram-facing connection.

Part of the SDK-migration rollout standardizing the Node starters onto `@deepgram/sdk`.

## How

- The `/api/live-text-to-speech` handler now opens the Deepgram side via `deepgram.speak.v1.createConnection({ model, encoding, sample_rate, container })`, wires `on('open'|'message'|'error'|'close')`, then `connect()` + `waitForOpen()`.
- Browser JSON control messages are mapped to the SDK's v1 speak socket methods: `Speak → sendText`, `Flush → sendFlush`, `Close → sendClose`. Messages that arrive before the socket opens are buffered and flushed on open.
- Deepgram → browser: binary audio (Blob/ArrayBuffer/Buffer) is forwarded as binary; JSON control objects are forwarded as JSON — the browser protocol is byte-for-byte unchanged, so **no frontend edits are required**.
- A single reused `DeepgramClient` resolves auth from `DEEPGRAM_API_KEY`; `DEEPGRAM_BASE_URL` is still honored via the SDK `environment` option.
- The `ws` browser server, JWT/session auth, `/api/session`, `/api/metadata`, and graceful shutdown are unchanged.
- Adds `@deepgram/sdk` `^5.7.0` to `package.json` (lockfile updated) and sets `deepgram.toml` `sdk = "@deepgram/sdk"`.

## Verified

- `node --check server.js` passes.
- `corepack pnpm install` updates the lockfile (`@deepgram/sdk 5.7.0`).
- Boot test: `DEEPGRAM_API_KEY=dummy PORT=8091 node server.js` starts cleanly and `GET /api/metadata` returns the expected JSON (with `sdk: "@deepgram/sdk"`).

## Manual verification needed before merge

- [ ] Live run with a real `DEEPGRAM_API_KEY`: connect over `/api/live-text-to-speech`, send `Speak`/`Flush`, and confirm streaming audio plays back end-to-end.
- [ ] Confirm the paired `*-html` frontend submodule works unchanged against this backend.

### CODE REVIEW

**Review by dg-coreylweathers — CHANGES_REQUESTED** [review]
**What this PR does**

It moves the live TTS connection in `server.js` onto `@deepgram/sdk` instead of a hand-rolled WebSocket. The audio tail-clipping problem from the August round is fixed and I verified it live, 3 runs of 3. One defect in the error path blocks merge.

**What I checked**

- **Is the tail clipping fixed?** Yes. Serializing the forwards through one promise chain resolves the reordering where `Flushed` overtook the final audio chunk. Verified live, 3 of 3.

- **Do upstream connection failures reach the browser as an error?** No. When Deepgram rejects the connection the browser receives zero messages and then a close with code 1000, which is the same signal as a healthy session ending.

  The SDK's `ReconnectingWebSocket` fires its synthetic `close` event *before* its `error` event on a failed handshake (`_handleError` calls `_disconnect()`, which calls `_handleClose(...)`, and only then the error listeners run). So the `close` handler at `server.js:279-284` shuts the browser socket before the `error` handler at `server.js:268-277` or the `waitForOpen` rejection at `server.js:330-340` can send anything.

  Expected: the PR body promises "On an upstream open failure, send a CONNECTION_FAILED Error frame before closing with 1011", and the pre-PR bridge on `main` sent a `PROVIDER_ERROR` frame with the message and then forwarded Deepgram's own close code.

  Observed, live in Docker across three failure modes: a fake key (Deepgram 401), `model=aura-bogus-en` (403 INSUFFICIENT_PERMISSIONS), and `container=wav` (400 UNSUPPORTED_AUDIO_FORMAT) each produced zero frames and close 1000. The bundled frontend shows "Disconnected", never "Error". A developer with a wrong key sees connect-then-drop with no explanation in the browser; the only clue is a server log line.

  Recommended fix, verified live on this branch. It reports pre-open failures from the close handler (which the SDK fires first), carries the SDK's status message (which contains no header or key), closes 1011, and routes the browser close through `sendChain` so no queued audio is dropped:

  ```js
  let lastDgError = null;
  dgSocket.on('error', (error) => {
    const message = error?.message ?? String(error);
    console.error('Deepgram socket error:', message);
    lastDgError = message;
    if (!dgReady) return; // pre-open failure is reported by the close handler below
    sendChain = sendChain.then(() => {
      if (clientWs.readyState === WebSocket.OPEN) {
        clientWs.send(JSON.stringify({
          type: 'Error',
          description: message || 'Deepgram connection error',
          code: 'PROVIDER_ERROR'
        }));
      }
    });
  });

  dgSocket.on('close', () => {
    console.log('Deepgram connection closed');
    sendChain = sendChain.then(() => {
      if (clientWs.readyState !== WebSocket.OPEN) return;
      if (!dgReady) {
        clientWs.send(JSON.stringify({
          type: 'Error',
          description: `Failed to connect to Deepgram TTS${lastDgError ? `: ${lastDgError}` : ''}`,
          code: 'CONNECTION_FAILED'
        }));
        clientWs.close(1011, 'Deepgram connection failed to open');
      } else {
        clientWs.close(1000, 'Deepgram connection closed');
      }
    });
  });
  ```

  With that applied, the fake-key run returned `{"type":"Error","description":"Failed to connect to Deepgram TTS: Unexpected server response: 401","code":"CONNECTION_FAILED"}` then close 1011, the bogus-model run returned the same shape with 403, and the ordering and browser-close scenarios still passed. The fake key appeared in neither the browser frames nor the server log.

**Also worth fixing**

- **SDK auto-reconnect can open a Deepgram connection the bridge can never use.** The SDK's `_handleClose` calls `_connect()` for any close code other than 1000, defaulting to 30 attempts with the first retry after 1 to 5 seconds, while the bridge has already closed the browser. Pass `reconnectAttempts: 0` to `deepgram.speak.v1.createConnection({...})` at `server.js:213`.
- **AGENTS.md describes the removed implementation and promises an option the API rejects.** It still says "WebSocket proxy" and "Modify these in the backend where the Deepgram URL is constructed", but no URL is constructed now. Its `container` row lists `none`, `wav`, `ogg`, while the live streaming endpoint answers `container=wav` with HTTP 400 `UNSUPPORTED_AUDIO_FORMAT` ("expected none or no container specified"). Point the customization guide at the `createConnection` call, add `@deepgram/sdk` to Dependencies, and list `none` only.

Smaller items: `server.js:215` logs the whole error object while the other handlers log `error?.message`; "Deepgram connection closed" logs twice per session because `V1Socket.close()` synthesizes a second close event; and `sample_rate` reaches the SDK as a string at `server.js:202` while `ConnectArgs.sample_rate` is typed as a number.

Note: merge #28 first, then rebase this branch's lockfile.

**Review by dg-coreylweathers — CHANGES_REQUESTED** [review]
**What this PR does**

It moves the server's Deepgram-facing WebSocket onto `@deepgram/sdk` (pinned to 5.7.0) instead of hand-written plumbing. Both defects from the earlier rounds are fixed and I verified both against the live API: the audio's last chunk no longer arrives after the `Flushed` "done" message, and a failed Deepgram connection now reaches the browser as an error frame with close code 1011 instead of a silent normal close. One documentation defect in the table this PR just edited blocks merge.

**What I checked**

- **Is the audio tail clipping fixed?** Yes. Three paired live runs: the bridge and a direct connection to `wss://api.deepgram.com/v1/speak` both returned 455 audio frames and exactly 1,894,738 bytes, with `Flushed` immediately after the last audio frame. Byte-identical 3/3.
- **Is the ordering fix load-bearing, or is the race imaginary?** Load-bearing. The live wire never delivered audio and `Flushed` in one batch, so a copy with the promise chain removed also passed 3/3. Forcing the condition with a fake upstream that writes 8 audio frames plus `Flushed` in one socket write: this branch orders them correctly 3/3; with the chain removed, `Flushed` lands at index 1 with 7 audio frames behind it, 3/3.
- **Does a failed upstream connection reach the browser as an error?** Yes, in all five failure modes I ran live — fake key (401), `container=wav` (400), `encoding=mp3` (400), `model=aura-bogus-en` (403), `sample_rate=abc` (400). Each delivered one `{"type":"Error","code":"CONNECTION_FAILED"}` frame then close 1011. No API key appeared in any frame or log line.
- **Does `reconnectAttempts: 0` actually stop the retry?** Yes. The SDK reads it as `reconnectAttempts ?? 30`, so 0 survives, and a fake upstream saw exactly one connection attempt.
- **Does the audio decode?** Yes. A live run returned 320,990 bytes; read back through a WAV header it is PCM format code 1, mono, 48000 Hz, 16-bit, 3.344 s, RMS 1893 (not silence).
- **Do `Clear` and `Close` work?** Yes, live — `Clear` mid-stream returned `Cleared` with 0 audio bytes after it, `Close` closed the browser socket with 1000. The PR body listed both as unverified.
- Also passing: `pnpm test` in `node:24` (1 test, 1 pass), `pnpm install --frozen-lockfile`, and `--frozen-lockfile` again after merging #28 into this branch locally, with `qs` resolving to 6.16.0.

**What to fix**

- **Blocking — `AGENTS.md:90-95` offers audio settings the streaming endpoint refuses.** Live against production: `encoding=mp3` and `encoding=opus` return HTTP 400 at 22050, 24000 and 48000, so neither works at any rate; `mulaw` and `alaw` return 400 at the table's own `48000` default (`mulaw` works at 8000/16000, `alaw` at 8000); and `sample_rate` is a fixed list — 8000/16000/24000/32000/44100/48000 work while 22050, 12345 and 96000 return 400 — not the `8000`-`48000` range the table claims. A developer following this table gets no audio and a bare "Unexpected server response: 400". Replace rows 93-94 with:

  ```
  | `encoding` | `linear16` | `linear16` (8000/16000/24000/32000/44100/48000 Hz), `mulaw` (8000/16000 Hz), `alaw` (8000 Hz) | Audio encoding |
  | `sample_rate` | `48000` | One of `8000`, `16000`, `24000`, `32000`, `44100`, `48000` — must be a rate the chosen encoding supports | Audio sample rate |
  ```

- **Should-fix — nothing tests the ordering fix.** The one test in `test/server.test.js:40` covers only the failed handshake, so deleting the promise chain at `server.js:271-276` leaves every gate green. Add a test using the same fake-upstream burst described above, asserting zero audio frames arrive after `Flushed`.
- **Should-fix — the new test runs in no gate.** `deepgram.toml:51` still says `command = "echo 'No tests configured'"`, `make test` runs the contracts script, and no workflow runs tests. Point `[test]` and the `Makefile` target at `corepack pnpm test`.
- **Should-fix — `package.json:40` pins `ws` at `8.18.0`**, which has an open high-severity memory-exhaustion advisory (patched 8.21.0) and an open medium uninitialized-memory advisory (patched 8.20.1), on the socket the browser connects to. This PR's lockfile already installs the patched `8.21.1` next to it for the SDK; pinning `8.21.1` fixes the app's own socket and collapses the duplicate.

Nits: a non-numeric `sample_rate` becomes the literal `NaN` on the request (`server.js:202`); the close handler's read of `lastDgError` works only because `sendChain.then(...)` defers it, which deserves a comment (`server.js:299-316`); an upstream failure after the socket opens closes the browser with 1000 while a pre-open failure closes 1011 (`server.js:282-297`); and 5.11.0 is the current SDK — I ran this branch against it and ordering, audio and failure reporting all behaved identically.

**What still needs you**

The Fly.io deploy workflow runs on every push to `main` and has failed on all four pushes since 2026-02-27 (last green: 2026-02-12). That is pre-existing, not this PR. The live app is up, not 502 — it returns 200 on `/`, `/api/metadata` and `/api/session` — but it is serving the 2026-02-12 build, so merging this will not put the SDK bridge in front of the public until the deploy is fixed. The logs have expired (HTTP 410) and the likely cause is the `FLY_API_TOKEN` secret, which cannot be checked from the repo. Also: merge #28 first — I verified the combined lockfile installs frozen in that order, whereas merging this first leaves #28's lockfile needing a regen.

**Comment by GregHolmes** [comment]
Corey — the branch now includes the requested streaming-format documentation correction, Blob-burst audio-before-Flushed regression coverage, runnable pnpm test wiring, and the direct ws security update. The current head is a33b42c. PR #28 remains the required merge prerequisite; after it merges, this branch will receive its final rebase and frozen-lockfile check. Please re-review the current changes when ready.

**Review by dg-coreylweathers — CHANGES_REQUESTED** [review]
**What this PR does**

It moves the server's Deepgram-facing WebSocket onto `@deepgram/sdk` (pinned 5.7.0) instead of hand-written plumbing, keeping the browser-facing contract the same. Everything from the two earlier rounds is fixed. The last commit changed the A-law sample rates I had asked for, and it is right and I was wrong: A-law streams at 16000 Hz as well as 8000 Hz.

**What I checked**

- **Does a wrong API key now reach the browser as an error?** Yes. Live against production with a bogus key: one `{"type":"Error","code":"CONNECTION_FAILED","description":"...Unexpected server response: 401"}` frame at 391 ms, then close 1011. The server log confirms the SDK still fires `close` before `error`, which is why reporting from the `close` handler is the thing that works.
- **Is that fix load-bearing, or does the failure report itself?** Load-bearing. With the pre-open branch removed, the same live wrong-key run returns zero frames and close 1000 — the silent disconnect from the September 11 round.
- **Is the audio tail still intact?** Yes. A live run returned 205 audio frames and 853,744 bytes, `Flushed` last, zero audio frames after it.
- **Does the new ordering test actually catch a regression?** Yes. Deleting the promise chain makes `test/server.test.js:166` fail with `'audio' !== 'Flushed'`.
- **Is the audio-format table true?** Yes, 23 live cases. Every combination it lists returns audio; every value it dropped returns HTTP 400 with the endpoint naming the reason. `alaw` at 8000 and 16000 Hz both return 71 frames.
- **Can the API key leak through an error path?** No. With a deliberately wrong 40-character key, the key appears in no browser frame and nowhere in the server log, and no `Authorization` header text is logged.
- **Do #27 and #28 collide on the lockfile?** No, in either order — both auto-merge, land `qs` at 6.16.0, and install frozen. #28-first is a preference, not a requirement.
- **How many checks actually ran?** Zero. `check-runs` on the head commit reports `total_count: 0` and the combined status has no contexts. There is no Snyk app here and no green mark; the merge box is blocked only by the earlier changes-requested review.
- Everything else passed: frozen install, `corepack pnpm test` (2 tests), `make test`, `make check`, `Clear`/`Close` live, buffered pre-open messages, no upstream connection leak on an early browser disconnect, and a repo-wide sweep finding no stale format claims outside `AGENTS.md`.

**What to fix before merge**

- **[B1] The contract conformance check now runs from no target at all.** `make test` replaced `bash contracts/tests/run-live-text-to-speech-app.sh` rather than gaining a sibling, and no target in the `Makefile`, `deepgram.toml` or `package.json` still references `contracts/tests`. Expected: `make test` runs the repo's tests and the contract conformance run stays reachable. Observed: on `main`, `make test` ran the contracts script behind `.env` and submodule guards; on this branch it runs `pnpm test` only, so the conformance script is unreachable. Fix: add a separate `make test-contracts` carrying the old guards. That replacement is what my last round asked for, so this is a consequence of my instruction, not a deviation from it.
- **[S1]** No `pull_request` workflow runs the new tests. The only workflows are the Fly deploy on pushes to `main` and a Context7 job on releases, so the burst-ordering test that exists to protect the audio fix runs only when someone types `make test`. A job running `corepack pnpm install --frozen-lockfile` then `corepack pnpm test` on `node:24` covers it.
- Four nits carried over, all in `server.js`: a non-numeric `sample_rate` still goes upstream as the literal `NaN`; the deferred `lastDgError` read still has no comment; a post-open failure closes 1000 while a pre-open one closes 1011; and 5.10.0 is the newest SDK this repo's own `.npmrc` age gate will install, not the 5.11.0 I named last round.

**What still needs you**

The Fly deploy has failed every push to `main` since 2026-02-27 (last green 2026-02-12). The public app returns 200, not 502, but it serves the February build, so merging will not put this bridge in front of visitors until the deploy is repaired. Separately, the `.npmrc` promises that packages must be 10 days old before install, but the pinned `pnpm@10.0.0` does not implement that setting — a 3-day-old package installs fine. Both are pre-existing and neither is what holds this PR; [B1] is.

**Comment by GregHolmes** [comment]
Corey - the final review feedback is addressed at `ebeef9e`.

- Restored the guarded contract target as `make test-contracts`; `make test` now runs the Node unit tests.
- Added the Node 24 pull-request test workflow; its first run passed.
- Validates numeric `sample_rate` before opening upstream, documents only supported streaming formats, disables unusable SDK reconnects, and covers handshake failure, post-open failure, browser Close, and burst audio ordering.

Please re-review the current head when ready.

**Review by dg-coreylweathers — CHANGES_REQUESTED** [review]
**What this PR does**

It moves the server's Deepgram-facing WebSocket onto `@deepgram/sdk` (pinned 5.7.0) instead of hand-written plumbing, keeping the browser-facing contract the same. Everything from all three earlier rounds is now fixed. Since the last round it restores the contract-conformance run as `make test-contracts`, adds a Node 24 pull-request test workflow, refuses a `sample_rate` that is not a positive whole number before calling Deepgram, and reports a mid-session upstream failure as `PROVIDER_ERROR` with close code 1011 while keeping 1000 for a normal end. #28 merged first and this branch is rebased onto it.

**What I checked**

- **Is the contract conformance run reachable again?** Yes. `make -n test-contracts` resolves to `bash contracts/tests/run-live-text-to-speech-app.sh`, both original guards are carried over verbatim, and a real run in a fresh clone stops at the `.env` preflight. `make test` runs the unit tests and `AGENTS.md` documents both.
- **Does a workflow actually run the tests now?** Yes. Run 35243334070 on the head commit passed in 14 seconds with pnpm 10.0.0 and Node 24.20.0, reporting 5 tests, 5 passed, 0 skipped. It is the first and only run of that workflow.
- **Are the three behavioral fixes load-bearing, or would the tests pass without them?** Load-bearing, all three. Deleting the promise chain fails the burst test with `actual: 'audio', expected: 'Flushed'`. Deleting the new post-open branch fails with `actual: 1000, expected: 1011`. Deleting the pre-open branch fails on the frame contents — and the live wrong-key run still closes 1011 rather than the old silent 1000, because the new post-open branch now covers that case too.
- **Does a wrong key still reach the browser as an error?** Yes. Live against production with a bogus key: one `Error`/`CONNECTION_FAILED` frame carrying "Unexpected server response: 401", then close 1011.
- **Is the audio tail still intact?** Yes. A live run returned 138 audio frames and 573,990 bytes with `Flushed` last and nothing after it.
- **Is a bad `sample_rate` refused locally now?** Yes. Live, `sample_rate=abc` returns `INVALID_REQUEST` and close 1008, and the server log shows no call to Deepgram at all.
- **Does the new `PROVIDER_ERROR` branch misfire on a normal session?** No. Browser `Close` still closes 1000 with no error frame, and a session left idle for 100 seconds after `Flush` was never closed by Deepgram.
- **Can the API key leak, or can a client kill the server?** No to both. The key appears in no frame and nowhere in the log. Hard-resetting the client TCP connection in the two windows where no error listener is attached leaves the process alive and still serving.
- Everything else passed: frozen install, `qs` at 6.16.0 with a single resolution, `ws` at 8.21.3 above both advisory floors, `make check`, and a repo-wide sweep finding no stale format or `make test` claims.

**What to fix before merge**

- **[S1] Two open high-severity advisories apply to the `toml` package this app loads at runtime.** `package.json` pins `toml` 3.0.0, from 2019. Expected: the direct dependencies in this manifest carry no open high-severity advisory, the standard this PR already applied to `ws`. Observed: GHSA-v5mp-jgw5-2x6j (prototype pollution, CVSS 8.2, patched in 4.1.2) and GHSA-82x6-q7mm-w9cf (uncontrolled recursion, CVSS 7.5, patched in 4.2.0) are both open against `package.json` and the lockfile. Neither is attacker-reachable here — `toml.parse()` runs once, at `server.js:147`, on the repo's own `deepgram.toml`, never on caller input — and both alerts predate this PR. Fix: `corepack pnpm update toml@5.0.0`. I confirmed it is a drop-in: 3.0.0, 4.2.0 and 5.0.0 all parse this repo's `deepgram.toml` to identical values, and 5.0.0 is old enough to clear the `.npmrc` age gate.

- **[S2] The error descriptions this PR now sends never appear on screen.** The pinned frontend reads neither `description` nor `code` on an `Error` frame: it sets the connection badge to the literal word "Error" and logs the object to the console, then the close handler replaces the badge with "Disconnected", and only close code 4401 is treated specially. So `CONNECTION_FAILED`, `PROVIDER_ERROR` and `INVALID_REQUEST` all look identical to a developer, and identical to clicking Disconnect. The server side is correct and is exactly what the earlier rounds asked for; the fix is in `live-text-to-speech-html` plus a submodule bump, so this can ship as its own PR rather than widening this one — your call which.

- Seven nits, details on request: the new `test` job is not a required status check and will never run on `main`; the workflow has no `permissions:` block; `@deepgram/sdk` is still 5.7.0 while 5.10.0 is the newest version the `.npmrc` age gate allows; `make test` has no `install` prerequisite so it fails on a fresh clone; the close handler is the one link in the send chain with no `.catch`; the `error` handler no longer forwards anything on its own; and a hexadecimal `sample_rate` passes the new integer check.

**What still needs you**

The Fly deploy has now failed five times running, last green 2026-02-12, and the newest failure is from merging #28 earlier today — so `main` is already red. The step names the cause: `no verified tokens; … verify: root banned`, meaning `FLY_API_TOKEN` needs replacing outside the repo. The public app answers 200 but still reports `"sdk":"N/A"`, so it is serving the February build and none of this bridge is live. Separately, `main` requires one approving review and zero status checks, so making the new `test` job required is a setting only an admin can flip. And the `.npmrc` still promises a 10-day package-age floor that the pinned `pnpm@10.0.0` does not implement.

For the record: there are no approving reviews on this PR, three changes-requested reviews from you, and no inline threads. Every claim in the latest author comment checks out against the commits and the CI run.

File: AGENTS.md [hunk 1]
```diff
@@ -85,14 +85,14 @@ Frontend: `cd frontend && corepack pnpm install`
...
+The WebSocket connection URL passes parameters to the backend. Update the defaults near the `deepgram.speak.v1.createConnection` call in `server.js`:
 
 | Parameter | Default | Options | Effect |
 |-----------|---------|---------|--------|
 | `model` | `aura-asteria-en` | Any aura-* voice | Voice selection |
-| `encoding` | `linear16` | `linear16`, `mp3`, `opus`, `mulaw`, `alaw` | Audio encoding |
-| `sample_rate` | `48000` | `8000`-`48000` | Audio sample rate |
-| `container` | `none` | `none`, `wav`, `ogg` | Audio container |
+| `encoding` | `linear16` | `linear16` (8000/16000/24000/32000/48000 Hz), `mulaw` (8000/16000 Hz), `alaw` (8000/16000 Hz) | Audio encoding |
+| `sample_rate` | `48000` | One of `8000`, `16000`, `24000`, `32000`, `48000` - must be a rate the chosen encoding supports | Audio sample rate |
```
Comment (dg-coreylweathers) [inline AGENTS.md:94]: This row and the `encoding` row above it dropped `44100` in this PR, and the live endpoint still accepts it. I sent `encoding=linear16&sample_rate=44100` to the streaming endpoint with `aura-asteria-en` twice: both runs returned 98 audio frames and 373,758 bytes, and that byte count divides out to the same 4.238 s of audio as the 48000 run, so the rate is genuinely honoured rather than resampled. `22050` does return HTTP 400, so the rest of the row is accurate.

Worth saying where the discrepancy actually lives: the SDK's `SpeakV1SampleRate` list also omits `44100` (while `SpeakV2SampleRate` includes it for `linear16`), so this edit is consistent with the SDK — the gap is on our side, not in this PR. My suggestion is to leave the table exactly as you have it, so the starter does not contradict the SDK version it pins, and let me chase the SDK list separately. The only ask is a line in the PR body noting the table changed, since it landed in a commit about the test command.

File: server.js [hunk 2]
```diff
@@ -131,102 +164,226 @@ app.get('/api/metadata', (req, res) => {
...
-    });
+  // Parse query parameters from the WebSocket URL
+  const url = new URL(request.url, `http://${request.headers.host}`);
+  const model = url.searchParams.get('model') || 'aura-asteria-en';
+  const encoding = url.searchParams.get('encoding') || 'linear16';
+  const sampleRateParam = url.searchParams.get('sample_rate');
+  const sampleRate = sampleRateParam === null ? 48000 : Number(sampleRateParam);
+  const container = url.searchParams.get('container') || 'none';
+
+  if (!Number.isInteger(sampleRate) || sampleRate <= 0) {
```
Comment (dg-coreylweathers) [inline server.js:206]: Nit, carried from earlier rounds. This catches the cases that matter, but `Number()` accepts some forms that are probably not intended: `Number("0x10")` is `16` and `Number("1e3")` is `1000`, both of which pass `Number.isInteger` and get forwarded upstream as a real rate. If you want only plain decimal input, a `/^\d+$/` test on the raw string before the `Number()` conversion would do it.

File: server.js [hunk 3]
```diff
@@ -131,102 +164,226 @@ app.get('/api/metadata', (req, res) => {
...
-      console.error('Client WebSocket error:', error);
-      if (deepgramWs.readyState === WebSocket.OPEN) {
-        deepgramWs.close();
+  dgSocket.on('close', (event) => {
+    if (dgClosed) return;
+    dgClosed = true;
+    console.log('Deepgram connection closed');
+    // The SDK emits close before error for failed handshakes and terminal errors.
+    // Reading lastDgError in this queued callback lets that later error listener run first.
+    sendChain = sendChain.then(() => {
```
Comment (dg-coreylweathers) [inline server.js:308]: Nit: this is the one link in the send chain without a `.catch`. The message handler above attaches one, so a throw inside this callback becomes an unhandled rejection instead of a log line.

File: server.js [hunk 4]
```diff
@@ -131,102 +164,226 @@ app.get('/api/metadata', (req, res) => {
...
+    let msg;
+    try {
+      msg = JSON.parse(raw.toString());
+    } catch {
+      console.warn('Ignoring non-JSON message from client');
+      return;
+    }
+    if (!msg || typeof msg !== 'object' || Array.isArray(msg) || typeof msg.type !== 'string') {
+      console.warn('Rejecting invalid client message');
       clientWs.send(JSON.stringify({
```
Comment (dg-coreylweathers) [inline server.js:341]: Nit: this `send` has no `readyState` check, unlike `forwardToBrowser`, which returns early unless the socket is open. Low risk in practice since we are inside the client's own message handler, but it is the one send in the file that does not follow that convention.

File: package.json [hunk 5]
```diff
@@ -30,12 +31,13 @@
   },
   "homepage": "https://github.com/deepgram-starters/node-live-text-to-speech#readme",
   "dependencies": {
+    "@deepgram/sdk": "5.7.0",
```
Comment (dg-coreylweathers) [inline package.json:34]: Nit: this is pinned to 5.7.0, and 5.10.0 is the newest release old enough to clear the package-age floor in `.npmrc`. I ran this branch against a newer SDK in an earlier round and ordering, audio and failure reporting all behaved identically, so this is housekeeping rather than a fix.

File: Makefile [hunk 6]
```diff
@@ -93,8 +94,13 @@ start-frontend:
 	@echo "==> Starting frontend on http://localhost:8080"
 	cd frontend && $(PNPM) run dev -- --port 8080 --no-open
 
-# Run contract conformance tests
+# Run unit tests
 test:
```
Comment (dg-coreylweathers) [inline Makefile:98]: Nit: this target has no `install` prerequisite, so on a fresh clone with no `node_modules` it fails rather than installing first. `init` and `check` both declare one.

File: .github/workflows/test.yml [hunk 7]
```diff
@@ -0,0 +1,19 @@
+name: Test
+
+on:
+  pull_request:
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
```
Comment (dg-coreylweathers) [inline .github/workflows/test.yml:8]: Nit: this workflow declares no `permissions:` block, so the job runs with the repository default token scope. `permissions: contents: read` is enough for what it does. Same applies to the other two workflows here, so it is pre-existing rather than introduced by this PR.

**Review by dg-coreylweathers — APPROVED** [review]
Approving. Both items from my last round are fixed, and I verified each against the live API rather than taking the summary.

## What this PR does

It moves the server's Deepgram-facing WebSocket onto `@deepgram/sdk` instead of hand-written plumbing, keeping the browser-facing message contract the same. Since the last round it upgrades `toml` past two open high-severity advisories, bumps the pinned frontend so the error text the server sends actually reaches the screen, refuses a client message that is not a JSON object with a string `type`, and scopes the test command to `test/`.

## What I checked

- **Are both `toml` advisories cleared?** Yes. 4.2.0 is the exact minimum that clears both of them (GHSA-v5mp-jgw5-2x6j is patched in 4.1.2, GHSA-82x6-q7mm-w9cf in 4.2.0). It parses this repo's own `deepgram.toml` to identical values, and it is old enough to clear the package-age floor in `.npmrc`.
- **Does toml's newer Node requirement break anything here?** No. It needs Node 20 or newer, and this repo's manifest, its CI job and both Dockerfile stages are all on 24.
- **Does a developer now see *why* a connection failed?** Yes. The frontend reads the `description` field off an `Error` frame and no longer replaces it with "Disconnected" when the socket closes, so a wrong key now reads "Failed to connect to Deepgram TTS: Unexpected server response: 401" instead of looking the same as clicking Disconnect.
- **Does the new message guard reject anything legitimate?** No. The frontend sends exactly two shapes, `{type:'Speak',text}` and `{type:'Flush'}`, and the guard runs after the JSON parse and before anything is buffered or forwarded upstream. A real `Close` still passes through and is still treated as a normal end.
- **Is that guard load-bearing?** Yes. With it removed, the new test fails: the client gets close code 1001 instead of 1008.
- **Do the gates pass?** Yes. 6 of 6 tests, a frozen install, `qs` resolving to 6.16.0 as a single copy, `ws` at 8.21.3. CI's own run on this commit reports the same 6 tests.
- **Can the API key leak through an error path?** No. With a deliberately wrong key, it appears in no browser frame and nowhere in the server log.

## One thing worth a follow-up

The audio-format table dropped `44100`, and the live endpoint still accepts it — details inline on `AGENTS.md`. Short version: the edit matches the SDK's own speak v1 sample-rate list, so the inconsistency is really on our side rather than in this PR, and I would leave the table as it is now rather than have the starter contradict the SDK it pins. I will follow up on the SDK list separately. The only ask here is that the change be mentioned in the PR body, since it currently rides inside a commit titled "scope live TTS test command".

One small thing: the PR body's verified list says 5 tests; the head runs 6.

The rest are nits left inline, none of them blocking.
